### PR TITLE
Typeck highlevel before bodies

### PR DESCRIPTION
**Pull request by pnkfelix**
typeck: Do high-level structural/signature checks before function body checks.

This avoids various ICEs, e.g. premature calls to cat_expr that yield the dreaded "cat_expr Errd" ICE.

However, it also means that some early error feedback is now not provided.  This may be for the best, because the error feedback were were providing in some of those cases were false positives -- it was spurious feedback and a distraction from the real problem.

So it is not 100% clear whether we actually want to put this change in or not.  I think its a net win, but others might disagree.

(Kudos to @arielb1 for suggesting this modification.)

### CODE REVIEW

File: src/test/compile-fail/associated-types-no-suitable-supertrait.rs [hunk 1]
```diff
@@ -17,13 +25,15 @@ trait Get : ::std::marker::MarkerTrait {
...
+    // (note that we no longer catch the error here, since the
+    //  error below aborts compilation.
+    //  See also associated-types-no-suitable-supertrait-2.rs
+    //  which checks that this error would be caught eventually.)
 }
 
 impl<T:Get> Other for T {
     fn uhoh<U:Get>(&self, foo: U, bar: <(T, U) as Get>::Value) {}
     //~^ ERROR the trait `Get` is not implemented for the type `(T, U)`
-    //~| ERROR the trait `Get` is not implemented for the type `(T, U)`
```
Comment (nagisa) [inline src/test/compile-fail/associated-types-no-suitable-supertrait.rs]: :+1: this commit just for this.

**Comment by ebfull** [comment]
Nice!

**Comment by kballard** [comment]
Conceptually, I like this change. I'd rather debug the high-level structural issues first, especially as issues in function bodies may actually just be symptoms of high-level issues.

**Comment by nikomatsakis** [comment]
r+

I've been thinking about doing exactly this, actually, so happy that you beat me to it. (The other thing I have been considering is adding some kind of rough dependency analysis so we can type-check callees before callers.)

**Comment by nikomatsakis** [comment]
Note though that we already have the `wf` pass, which is supposed to be *exactly this* distinction of signatures vs bodies. It was introduced because forcing types to have the right annotations avoided a lot of spurious errors. Note sure if it makes sense to keep the "three-phases". I think i'm inclined to r+ it as is but gradually refactor `check/mod` to break it up and make the pass structure clearer.

**Comment by pnkfelix** [comment]
normally we would not (should not) accept a change like this for beta, but, since this (1.) fixes ICEs (2.) is a user-experience improvement, and (3.) is pretty low-risk, we are taking it just because 1.0 release is a special case

going from nominated to (nominated, accepted)

**Pull request merged**